README.md
# homebridge-hftco2

![Tavis CI build status](https://travis-ci.com/chk-code/homebridge-hftco2.svg?branch=main&status=passed)

[HomeBridge](http://github.com/nfarina/homebridge) module for the DIY [HFT Stuttgart](https://www.hft-stuttgart.de/) [CO2 Ampel](https://www.hft-stuttgart.de/forschung/news/co2-ampel-lueften-gegen-covid-19).

It can be used to see the status of your own sensor in HomeKit.

## Features

- See Co2 sensor data (in ppm), temperature and humidity in HomeKit and the Home app
- Get the data from a sensor on your local network
- See history for temperature, etc. using the Elgato Eve app 
- Supported sensors / combinations : 
  - Sensirion SCD30

## Setup

First follow the instructions for [HomeBridge](http://github.com/nfarina/homebridge). 

Install `homebridge-hftco2` using `(sudo) npm install -g homebridge-hftco2`.

Configure your CO2 Ampel in the `homebridge` settings file. See [config.sample.json](config.sample.json). All settings except 'sensorID', `update_interval` (default 60 seconds) and `co2warnlevel` (default 800 ppm) are required.

See the documentation of the [fakegato-history](https://github.com/simont77/fakegato-history/blob/master/README.md#history-persistence) module for the list of supported options.

### The sensor & homebridge in your (home) network

To do this set `json_url` to

`http://<YOUR_CO2AMPEL_IP-ADDRESS>//data.json`

using the same local network as your sensor. 

*Note:* If you use a HomeKit hub (like a Apple TV or a iPad) you can still access your data from outside your home. See [Apple's documentation for details](https://support.apple.com/en-us/HT207057).


## Demo

![homebridge-hftco2 in apple home app](img/screenshot.jpeg)

CHANGELOG.md
# Changelog

## v1.0.2 - 2021-03-17
- [FIX] Bug in platform.js

## v1.0.2 - 2021-03-15
- [FIX] npm package issues

## v1.0.1 - 2021-03-14
- [FIX] First fixes 

## v1.0.0 - 2021-03-14
- [NEW] Init release

src/data_cache.js
const http = require('http');

class DataCache {
  constructor() {
    this.timestamp = null;
    this.temperature = null;
    this.humidity = null;
    this.co2 = null;;
  }

  updateFromURLSensor(url, callback) {
    this._loadCurrentSensorData(url, (error, json) => {
      if (error) {
        callback(error);
        return;
      }
      this._updateHumidity(json);
      this._updateTemperature(json);
      this._updateTimestamp(json);
      this._updateCO2(json);
      callback(null);
    })
  }

  _updateHumidity(json) {
    const humidityKeys = ['humidity'];
    for (let key of humidityKeys) {
      const value = this._findValue(json, key);
      if (value) {
        this.humidity = parseFloat(value);
        break;
      }
    }
  }
  _updateTimestamp(json) {
    const tsKeys = ['timestamp'];
    for (let key of tsKeys) {
      const value = this._findValue(json, key);
      if (value) {
        this.timestamp = parseFloat(value);
        break;
      }
    }
  }
  _updateTemperature(json) {
    const tempKeys = ['temperature'];
    for (let key of tempKeys) {
      const value = this._findValue(json, key);
      if (value) {
        this.temperature = value;
        break;
      }
    }
  }

  _updateCO2(json) {
    const co2Keys = ['co2'];
    for (let key of co2Keys) {
      const value = this._findValue(json, key);
      if (value) {
        this.co2 = parseFloat(value);
        break;
      }
    }
  }

  _findValue(json, key) {
    let basedata = json;
    // If loading data from API the result sometimes is
    // an array
    if (Array.isArray(basedata)) {
      basedata = basedata[0];
    }
    if (!basedata) {
      return null;
    }
    const sensorValues = basedata["sensordatavalues"];
    if (!sensorValues) {
      return null;
    }
    for (let valueSet of sensorValues) {
      if (key == valueSet["value_type"]) {
        return valueSet["value"];
      }
    }
    return null;
  }

  _loadCurrentSensorData(jsonURL, callback) {
    http.get(jsonURL, (res) => {
      const { statusCode } = res;
      const contentType = res.headers['content-type'];

      let error;
      if (statusCode !== 200) {
        error = new Error('Request Failed.\n' +
                          `Status Code: ${statusCode}`);
      } else if (!/^application\/json/.test(contentType)) {
        error = new Error('Invalid content-type.\n' +
                          `Expected application/json but received ${contentType}`);
      }
      if (error) {
        res.resume();
        callback(error, null);
        return;
      }

      res.setEncoding('utf8');
      let rawData = '';
      res.on('data', (chunk) => { rawData += chunk; });
      res.on('end', () => {
        try {
          const parsedData = JSON.parse(rawData);
          callback(null, parsedData);
        } catch (error) {
          callback(error, null);
        }
      });
    }).on('error', (error) => {
      callback(error, null);
    });
  }
}

module.exports = DataCache;

src/CO2AmpelAccessory.js
'use strict';

const Logger = require('./logger.js');
const DataCache = require('./data_cache.js');

class CO2AmpelAccessory {

  constructor (api, accessory, accessories, FakeGatoHistoryService, co2sensor) {

    this.api = api;
    this.accessory = accessory;
    this.accessories = accessories;
    this.FakeGatoHistoryService = FakeGatoHistoryService;

    this.co2sensor = co2sensor;

    this.dataCache = new DataCache();
    this.isUpdating = false;

    this.isCO2Detected = 0;

  // Update CO2 value by interval
	setInterval(() => {
		this.updateCache();
	}, this.co2sensor.update_interval * 1000);

  this.getService();

  }

  //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
  // Services
  //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

  async getService () {

    let service = this.accessory.getService(this.api.hap.Service.CarbonDioxideSensor);
    this.historyService = new this.FakeGatoHistoryService('room', this.accessory, {storage:'fs', path: this.api.user.storagePath(), disableTimer:true});

    if(!service){
      Logger.info('Adding CarbonDioxideSensor service', this.accessory.displayName);
      service = this.accessory.addService(this.api.hap.Service.CarbonDioxideSensor, this.accessory.displayName, this.accessory.context.config.type);
    }

    service
      .getCharacteristic(this.api.hap.Characteristic.CarbonDioxideDetected)
      .on('get', this.getCo2Detected.bind(this));
    service
      .getCharacteristic(Characteristic.CarbonDioxideLevel)
      .on('get', this.getCo2Level.bind(this));
  	// Set init CO2 value
    this.getCo2Level((err, value) => this
      .service
      .setCharacteristic(this.api.hap.Characteristic.CarbonDioxideLevel, value),
    );
    // Set init CO2 threshold
    this.service.setCharacteristic(this.api.hap.Characteristic.CarbonDioxideDetected, this.isCO2Detected);
  };

   refreshHistory = (callback) => { 
    
    await timeout(5000);
    
    this.historyService.addEntry({
      time: moment().unix(), 
      temp: this.dataCache.temperature, 
      humidity: this.dataCache.humidity, 
      ppm: this.dataCache.co2
    });
    
    setTimeout(() => {
      this.refreshHistory();
    }, 10 * 60 * 1000);

  };

  setCo2Detected = function (value) {
    if (value >= this.co2sensor.co2warnlevel) {
      this.isCO2Detected = 1;
    } else {
      this.isCO2Detected = 0;
    }
    this.service.getCharacteristic(this.api.hap.Characteristic.CarbonDioxideDetected).updateValue(this.isCO2Detected);
  };

  getCo2Detected = function (callback) {
    Logger.info(`CO2 detected: ${this.isCO2Detected}.`);
    callback(null, this.isCO2Detected);
  };
  getCo2Level = function (callback) {
    let co2level = this.co2sensor.co2
    callback(null, co2level);
  };

  updateCache = (callback) => {
    if (this.isUpdating) {
      if (callback) {
        callback(null);
      }
      return;
    }
    this.isUpdating = true;
    const updateCallback = (error) => {
      this.isUpdating = false;
      if (error) {
        Logger.error(`Could not get sensor data: ${error}`);
      }
      else {
        this.updateServices(this.dataCache);
      }
      if (callback) {
        callback(error);
      }
    };
    if (this.co2sensor.jsonURL) {
      this.dataCache.updateFromURLSensor(this.co2sensor.jsonURL, updateCallback);
      this.refreshHistory();
    }
  };
  updateServices = (dataCache) => {
    this.dataCache = dataCache;
    const { 
      co2,
      temperature, 
      humidity,
      timestamp,
    } = dataCache;
    this.setCo2Detected(co2);
    this.service.setCharacteristic(this.api.hap.Characteristic.CarbonDioxideLevel, co2);
  };
}

module.exports = CO2AmpelAccessory;

src/platform.js
'use strict';

const Logger = require('./logger.js');
const packageVer = require('../package.json');

//Accessories
const CO2AmpelAccessory = require('./CO2AmpelAccessory.js')

//CONSTANTS
const PLUGIN_NAME = 'homebridge-hftco2';
const PLATFORM_NAME = 'CO2AmpelPlatform';

var Accessory, UUIDGen, FakeGatoHistoryService;

module.exports = function (homebridge) {

  Accessory = homebridge.platformAccessory;
  UUIDGen = homebridge.hap.uuid;
  
  return CO2AmpelPlatform;

};

function CO2AmpelPlatform (log, config, api) {
  
  if (!api||!config) 
    return;

  Logger.init(log, true);
  FakeGatoHistoryService = require('fakegato-history')(api);

  this.api = api;
  this.accessories = [];
  this.config = config;
  
  this.devices = new Map();

  if(this.config.co2sensor && this.config.co2sensor.length) {
  
    this.config.co2sensor.forEach(co2sensor => {
    
      let error = false;

      if (!co2sensor.name) {
        Logger.warn('One of the CO2 Ampeln has no name configured. This device will be skipped.');
        error = true;
      } else if (!co2sensor.json_url) {
        Logger.warn('There is no JSON URL configured for this CO2 Ampel. This device will be skipped.', co2sensor.name);
        error = true;
      }

      if (!error) {
        
        let options = {
          charset: 'utf-8',
          language: 'en-us',
          version: '2.0'
        };
      
        const uuidCO2Ampel = UUIDGen.generate(co2sensor.name);
        
        if (this.devices.has(uuidCO2Ampel)) {
     
          Logger.warn('Multiple devices are configured with this name. Duplicate devices will be skipped.', co2sensor.name);
     
        } else {
          
          co2sensor.type = 'CarbonDioxideSensor';
          co2sensor.update_interval = Number.isInteger(co2sensor.update_interval) 
            ?  co2sensor.update_interval < 1 
                 ? 1 
                 : co2sensor.update_interval
            :  60;
          co2sensor.co2warnlevel = Number.isInteger(co2sensor.co2warnlevel) 
            ?  co2sensor.co2warnlevel < 1 
                 ? 1 
                 : co2sensor.co2warnlevel
            :  800;
          this.devices.set(uuidCO2Ampel, co2sensor);
          
        }
    
      }
      
    });
    
  }

  this.api.on('didFinishLaunching', this.didFinishLaunching.bind(this));
  
}

CO2AmpelPlatform.prototype = {

  didFinishLaunching: async function(){

    for (const entry of this.devices.entries()) {
    
      let uuid = entry[0];
      let device = entry[1];
      
      const cachedAccessory = this.accessories.find(curAcc => curAcc.UUID === uuid);
      
      if (!cachedAccessory) {
      
        const accessory = new Accessory(device.name, uuid);

        Logger.info('Configuring accessory...', accessory.displayName); 
        
        this.setupAccessory(accessory, device);
        
        Logger.info('Configured!', accessory.displayName);
        
        this.api.registerPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [accessory]);
        
        this.accessories.push(accessory);
        
      }
      
    }

    this.accessories.forEach(accessory => {
    
      const device = this.devices.get(accessory.UUID);
      
      try {
      
        if (!device)
          this.removeAccessory(accessory);
    
      } catch(err) {

        Logger.info('Device already removed? Skip removing.');
        Logger.debug(err);
     
      }
      
    });
  
  },
  
  setupAccessory: function(accessory, device){
    
    accessory.on('identify', () => {
      Logger.info('Identify requested.', accessory.displayName);
    });
    
    const manufacturer = 'HFT Stuttgart';
      
    const model = 'CO2 Ampel';
    
    const sensorID = device.sensorID
      ? device.sensorID 
      : 'SerialNumber';
    
    const AccessoryInformation = accessory.getService(this.api.hap.Service.AccessoryInformation);
    
    AccessoryInformation
      .setCharacteristic(this.api.hap.Characteristic.Manufacturer, manufacturer)
      .setCharacteristic(this.api.hap.Characteristic.Model, model)
      .setCharacteristic(this.api.hap.Characteristic.SerialNumber, sensorID)
      .setCharacteristic(this.api.hap.Characteristic.FirmwareRevision, packageVer.version);
      
    let co2sensor = device.co2sensor;
    
    delete device.co2sensor;
    
    accessory.context.config = device;
    
    new CO2AmpelAccessory(this.api, accessory, this.accessories, FakeGatoHistoryService, co2sensor);
    
    return;

  },

  configureAccessory: async function(accessory){

    const device = this.devices.get(accessory.UUID);

    if (device){
      Logger.info('Configuring accessory...', accessory.displayName);                                                                                            
      this.setupAccessory(accessory, device);
    }
    
    this.accessories.push(accessory);
  
  },
  
  removeAccessory: function(accessory) {
  
    Logger.info('Removing accessory...', accessory.displayName);
    
    let accessories = this.accessories.map( cachedAccessory => {
      if(cachedAccessory.displayName !== accessory.displayName){
        return cachedAccessory;
      }
    });
    
    this.accessories = accessories.filter(function (el) {
      return el != null;
    });

    this.api.unregisterPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [accessory]);
  
  }

};